text/1028-cli-hooks.rst

```diff
@@ -122,23 +125,105 @@ means that the hooks associated with applying schema changes must be executed
+also later introduce a ``[watch.gel-config]`` sub-table for monitoring changes
+to the various database config values and responding to them.
+
+The general structure of watch tables is going to use the key (potentially
+quoted) to specify *what is being watched* and the corresponding value to
+specify the script that will be triggered by changes.
+
+Since the triggered scripts will be executed in response to changes, we cannot
+meaningfully display the output of the scripts to the user. We may want to
+setup an optional logfile to record the script output. This can be specified
+in the general ``[watch]`` section as ``logfile="<path-to-logfile>"``.
+
+We may want to setup a bit of a delay before triggering the scripts so as to
+allow a sequence of changes to happen before the response script is triggered.
+This is mostly to reduce unnecessary multiple triggers for the same watched
+entity.
+
+Another consequence of executing triggered scripts in the background is that
+sometimes the changes are invalid in some way and the script will fail. This
+is considered part of normal operation (e.g. syntax error in a saved query
+file) and the scripts should be such that failure does not create some
+non-recoverable state.
+
+
+Files
+-----
+
+The ``[watch.files]`` table should contain keys that are interpreted as file system paths. The following special symbols should be allowed:
+
+- ``?`` to denote a single character wild card
+
+- ``*`` to denote an arbitrary string of characters in a name
+
+- ``**`` to denote an arbitrary sub-path
+
+The paths should be valid in the underlying file system (therefore using ``/``
+for Linux and ``\`` for Windows, etc.). The relative file paths are assumed to
+start at the project root.
+
+The values corresponding to the keys are strings that are going to be executed
+as shell commands, much like for the hooks.
+
+All watch scripts will use the project root directory as the execution
+directory. They are executed using ``/bin/sh`` on all platforms. On Windows,
+the scripts are always executed in WSL.
+
+An example of this configuration::
+
+    [watch.files]
+    "queries/*.edgeql"="npx @edgedb/generate queries"
+
+
+CLI Interactions
+----------------
+
+Currently ``gel watch`` is not relying on the ``[watch.files]`` config, but
+going forward it would be nice to generalize it using the watch mechanism
+instead of having special treatment for it. We may need to introduce a special
+migration command though that mimics the behavior of ``gel watch`` in terms of
+applying schema changes without writing a migration step for each such change.
+Perhaps a ``gel migration try`` command that would attempt to apply changes
+based on the current schema files without writing a corresponding migration
+file.
+
+With that CLI change we could make ``gel watch`` command create the following
+entry in ``gel.toml``::
+
+    [watch.files]
+    "dbschema/**.schema"="gel migration try"
```

Comment:
```suggestion
    "dbschema/**.gel"="gel migration try"
```

I kinda like the generalization idea but I think it's a step too far. IMO `gel watch` should just require either or both `--schema` / `--files` flags and do the right thing.

Reply:
From reading the code, I see that, currently, `gel watch` is triggering `crate::migrations::dev_mode::migrate`, without having any state in-between invocations, apart of error reporting.

https://github.com/edgedb/edgedb-cli/blob/master/src/watch/main.rs#L154-L182

So I actually do think generalization is possible.